src/utils/mempool.rs

```diff
@@ -138,3 +145,54 @@ impl io::Seek for MemPool {
         io::Seek::seek(&mut self.file, pos)
     }
 }
+
+fn create_shm_fd() -> io::Result<RawFd> {
+    loop {
+        match memfd::memfd_create(
+            CStr::from_bytes_with_nul(b"smithay-client-toolkit\0").unwrap(),
+            memfd::MemFdCreateFlag::MFD_CLOEXEC,
+        ) {
+            Ok(fd) => return Ok(fd),
+            Err(nix::Error::Sys(Errno::EINTR)) => continue,
+            Err(nix::Error::Sys(errno)) => return Err(io::Error::from(errno)),
+            Err(_) => break,
+        }
+    }
+
+    // Fallback to using shm_open
+    let mut rng = thread_rng();
+    let mut mem_file_handle = format!(
+        "/smithay-client-toolkit-{}",
+        rng.gen_range(0, ::std::u32::MAX)
+    );
+    loop {
+        match mman::shm_open(
+            mem_file_handle.as_str(),
+            fcntl::OFlag::O_CREAT
+                | fcntl::OFlag::O_EXCL
+                | fcntl::OFlag::O_RDWR
+                | fcntl::OFlag::O_CLOEXEC,
+            stat::Mode::S_IRUSR | stat::Mode::S_IWUSR,
+        ) {
+            Ok(fd) => match mman::shm_unlink(mem_file_handle.as_str()) {
+                Ok(_) => return Ok(fd),
+                Err(nix::Error::Sys(errno)) => {
+                    unistd::close(fd);
+                    return Err(io::Error::from(errno));
+                }
+                Err(err) => panic!(err),
+            },
+            Err(nix::Error::Sys(Errno::EEXIST)) => {
+                // If a file with that handle exists then change the handle
+                mem_file_handle = format!(
+                    "/smithay-client-toolkit-{}",
+                    rng.gen_range(0, ::std::u32::MAX)
+                );
+                continue;
+            }
+            Err(nix::Error::Sys(Errno::EINTR)) => continue,
+            Err(nix::Error::Sys(errno)) => return Err(io::Error::from(errno)),
+            Err(err) => panic!(err),
```

Comment:
same here, please use `unreachable!(err)`

Reply:
This branch (and the similar one on the other match) covers nix-specific errors that supposedly cannot occur in our case. As such I'd prefer to use the `unreachable!(err)` macro instead of a regular `panic!(err)`, to better convey this.

Reply:
https://github.com/Smithay/client-toolkit/pull/26/commits/978b03742abdaaac18512fa3218efb9be51effbb should do this :)